caddyhttp/digestauth/digest.go

```diff
@@ -0,0 +1,296 @@
+/*
+Package digest provides HTTP Digest authentication.
+
+You may use Apache style htdigest files,
+a simple map of userXrealm->MD5(user:realm:password) or your own implementation of a user
+lookup function (if perhaps you use a database).
+
+The API is designed for HTTP request routers like Martini. It provides a handler, which
+will do nothing to the HTTP response for valid authentication. For invalid authentication
+it will make an http.Error(), which will usually be a 401 (Unauthorized) to request
+authentication from a user.
+
+You might use digest like this if you were getting passwords from your own source:
+
+     m := martini.Classic()
+     ...
+     myUserStore := auth.NewSimpleUserStore(  map[string]string{
+			"foo:mortal": "3791e8e14a10b3666ba15d9e78e4b359",    // pw is 'bar'
+			"Mufasa:testrealm@host.com": "939e7578ed9e3c518a452acee763bce9",   // pw is 'Circle Of Life'
+                    })
+     ...
+     digester := auth.NewDigestHandler( "mortal", nil, nil, myUserStore )
+     m.Use( digester.ServeHTTP )   // this will force authentication of all requests, you can be more specific.
+
+If you want to use htdigest files (because I know you thought about passing credentials on the command line
+or in environment variables and you should know better):
+
+     m := martini.Classic()
+     ...
+     // Read file: hint, the nil is standing in for a malformed line reporter function.
+     myUserFile,err := auth.NewHtdigestUserStore("path/to/my/htdigest/file", nil)
+     if err != nil {
+		log.Fatalf("Unable to load password file '%s': %s", digestfile, err.Error())
+     }
+     ...
+     digester := auth.NewDigestHandler( "My Realm", nil, nil, myUserFile )
+     ...
+     m.Post("/my-sensitive-uri", digester.ServeHTTP, mySensitiveHandler)  // just protect this one, notice chained handlers.
+
+You will have noticed that both New*Handler calls included a pair of
+nils. If you don't like the way nonces are created, tracked, and expired,
+then you will want to replace one or both of these with your own
+implementations of Nonce and NonceStore.  Otherwise, ignore those
```

Comment:
More idiomatic is to not take optional parameters in constructors and make them exported fields instead that users can optionally set to be non-nil before use.